Refactor air alarm TLV and modes.

vincentiusvin opened the pull request:
## About The Pull Request
#65902 but with a lot of fat trimmed.
Ditched the lazy initialization thing since the memory gains isn't worth the increased code complexity.
Ditched the vent bitflag changes, that will come in the future maybe.
Ditched the thermomachine and alarm resubtyping. Will also maybe come in the future.

Added the operating mode datums so we can reduce the amount of code in the main air alarm page. The datums lets us write tidier code too.

Scope is much much more limited than the old PR. The diffs are mostly just me splitting and moving the files around. Things to pay attention to are:
- `_air_alarm.dm`'s UI procs and things that deals with TLV and modes.
- `air_alarm_variants.dm`'s coldroom and server room numbers
- `air_alarm_thresholds.dm`
- `air_alarm_modes.dm`

Smaller changes on:
- `air_alarm_interact.dm` should be untouched except for one select_mode call.
- `air_alarm_circuit.dm`'s also untouched except the part that deals with TLV and modes
- `wires/air_alarm.dm`'s also untouched except for the panic siphon wire.

<details>
<summary>Images</summary>
<img src="https://user-images.githubusercontent.com/54709710/209765698-47bd2564-cc6a-4829-9418-97d8814d9f7a.png"/>
<img src="https://user-images.githubusercontent.com/54709710/209765764-7f59fa2d-66ad-4bc6-8687-a1024898823c.png"/>
<img src="https://user-images.githubusercontent.com/54709710/210332278-17d286fe-6d78-44f2-ac6f-77deb0dac92e.png"/>
</details>

## Why It's Good For The Game
Cleaner code for the backend, static typing for the frontend, bugfixes.
New UI for the threshold thing.

## Changelog
:cl:
qol: made the threshold page of air alarm use colloquial terms instead of code terms. Added reset and disable buttons.
qol: Air alarm's gas breakdown now displays the moles, percentage, and partial pressure.
fix: operating mode selection in air alarm will be red if the selected mode is dangerous.
fix: fixed the area atmosphere alarm button on air alarm not working.
code: told the game that hyper nob isn't dangerous.
refactor: refactored the threshold values and operating mode of air alarms in the backend. No changes expected except the ones stated above.
/:cl:

vincentiusvin committed:
Move things around

vincentiusvin committed:
Basic implementation

vincentiusvin committed:
Implemented variants

vincentiusvin committed:
Got rid of the duplicate armor datums

vincentiusvin committed:
Renames galore

vincentiusvin committed:
More moving

vincentiusvin committed:
Clean ui_data

vincentiusvin committed:
TS implementation

vincentiusvin committed:
UI improvements

vincentiusvin committed:
Shuffled procs around

vincentiusvin committed:
Refactored the modes into datums

vincentiusvin committed:
Frontend mode refactor

vincentiusvin committed:
Naughty naughty moth

vincentiusvin committed:
Bugfixes

vincentiusvin committed:
Fixes

vincentiusvin committed:
Bugfixes

vincentiusvin committed:
Documentation and settling old debt

vincentiusvin committed:
Woops

vincentiusvin committed:
Vin you bumbling fool

Fikou committed:
Merge branch 'master' into alarm

vincentiusvin committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into alarm

LemonInTheDark commented:
Classic tomorrow problem

LemonInTheDark commented on code/modules/atmospherics/machinery/air_alarm/_air_alarm.dm:
```diff
@@ -0,0 +1,529 @@
+	var/locked = TRUE
+	var/aidisabled = 0
+	var/shorted = 0
+	var/buildstage = AIRALARM_BUILD_COMPLETE // 2 = complete, 1 = no wires,  0 = circuit gone
```
similar

LemonInTheDark commented on code/modules/atmospherics/machinery/air_alarm/_air_alarm.dm:
```diff
@@ -0,0 +1,529 @@
+		name = "[get_area_name(src)] Air Alarm"
+
+	tlv_collection = list()
+	tlv_collection["pressure"] = new /datum/tlv/pressure
```
part of me is like "We could make these immutable!" but I bet it's like no cost/memory overhead, since they're just datums and rely on just init vars.
maybe better this way

vincentiusvin replied:
True. Not really worth the extra time coding too.

LemonInTheDark commented on code/modules/atmospherics/machinery/air_alarm/_air_alarm.dm:
```diff
@@ -0,0 +1,529 @@
+		else
+			tlv_collection[gas_path] = new /datum/tlv/no_checks
+
+	selected_mode = GLOB.air_alarm_modes[/datum/air_alarm_mode/filtering]
```
might be better to use a getter for this. makes their status a bit more clear, hopefully prevents people new()ing them

vincentiusvin replied:
already made a setter but forgot to use it here. Hope thats fine.

LemonInTheDark replied:
that'd be good enough, sure

LemonInTheDark commented on code/modules/atmospherics/machinery/air_alarm/_air_alarm.dm:
```diff
@@ -0,0 +1,529 @@
+#define AALARM_REPORT_TIMEOUT 100
```
is this why it's _air_alarm and not just air_alarm

vincentiusvin replied:
no i just like having the most important file show up at the top.

LemonInTheDark replied:
also if this is local undef it at the end

vincentiusvin replied:
ok

vincentiusvin replied:
lmao its actually unused.

LemonInTheDark commented on code/modules/atmospherics/machinery/air_alarm/_air_alarm.dm:
```diff
@@ -0,0 +1,529 @@
+				"name" = "Temperature",
+				"unit" = "K",
+			)
+		else
+			singular_tlv += list(
+				"name" = GLOB.meta_gas_info[threshold][META_GAS_NAME],
+				"unit" = "kPa",
+			)
```
so the assumption here is gas is the only acceptable third option? could you be explicit about that?

LemonInTheDark replied:
also why do the list() thing? guess it's a bit easier to read then setting each var individually but I've not seen it before

vincentiusvin replied:
> also why do the list() thing?
oh shit right thats kinda dumb

LemonInTheDark replied:
why not a final else if instead of a trailing else? oh right the thing's a fucking gas id god damn it.
you mind leaving a comment on it by the else?

vincentiusvin replied:
yeah sure

vincentiusvin replied:
Nah its a gaspath, we can totally do `in subtypesof()` if you want that extra safety.

LemonInTheDark replied:
it's ok, just comment it explicitly so readers "get" it

LemonInTheDark replied:
that or just check the truthyness of meta_gas_info
yeah do that I like that better

LemonInTheDark commented on code/modules/atmospherics/machinery/air_alarm/_air_alarm.dm:
```diff
@@ -0,0 +1,529 @@
+		return
+
+	var/mob/user = usr
+	var/area/area = get_area(src)
+	ASSERT(!isnull(area))
```
mothposting

LemonInTheDark commented on code/modules/atmospherics/machinery/air_alarm/_air_alarm.dm:
```diff
@@ -0,0 +1,529 @@
+
+	var/ref = params["ref"]
+
+	// Possible machines this can refer to
+	var/obj/machinery/atmospherics/components/unary/vent_pump/vent = isnull(ref) ? null : locate(ref) in area.air_vents
+	var/obj/machinery/atmospherics/components/unary/vent_scrubber/scrubber = isnull(ref) ? null : locate(ref) in area.air_scrubbers
```
is locate(null) nonvalid?

LemonInTheDark replied:
also you mind just doing an if wrap if so this is too much for one line, specially with the in

vincentiusvin replied:
Its probably an optimization.

LemonInTheDark commented on code/modules/atmospherics/machinery/air_alarm/_air_alarm.dm:
```diff
@@ -0,0 +1,529 @@
+
+	switch (action)
+		if ("power")
+			if (!isnull(vent))
+				vent.on = !!params["val"]
+				vent.update_appearance(UPDATE_ICON)
+				vent.check_atmos_process()
+			else if (!isnull(scrubber))
+				scrubber.on = !!params["val"]
+				scrubber.update_appearance(UPDATE_ICON)
+				scrubber.check_atmos_process()
```
```suggestion
			var/obj/machinery/atmospherics/components/powering = vent || scrubber
			powering .on = !!params["val"]
			powering .update_appearance(UPDATE_ICON)
			powering .check_atmos_process()
```

LemonInTheDark commented on code/modules/atmospherics/machinery/air_alarm/_air_alarm.dm:
```diff
@@ -0,0 +1,529 @@
+			if (isnull(vent))
+				return TRUE
+
+			var/new_checks = clamp((text2num(params["val"]) || 0) ^ ATMOS_INTERNAL_BOUND, NONE, ATMOS_BOUND_MAX)
```
how often do you do this clamping? could it be made into a macro?

vincentiusvin replied:
Tidying vents internal was in my original PR but I dropped it, wanna keep this atomic.

LemonInTheDark commented on code/modules/atmospherics/machinery/air_alarm/_air_alarm.dm:
```diff
@@ -0,0 +1,529 @@
+
+	var/area/our_area = get_area(src)
+	var/color
+	switch(max(danger_level, !!our_area.active_alarms[ALARM_ATMOS]))
```
this is fuckin weird

LemonInTheDark replied:
I know it already exists but man s weird

LemonInTheDark commented on code/modules/atmospherics/machinery/air_alarm/_air_alarm.dm:
```diff
@@ -0,0 +1,529 @@
+
+	var/area/our_area = get_area(src)
+	var/state
+	switch(max(danger_level, !!our_area.active_alarms[ALARM_ATMOS]))
```
think we should just make it a proc? oh but that'd do almost nothing. guess it would be useful to pull out just to name it, and for reuse in future

vincentiusvin replied:
Ill try and detach danger_level from area code and keep it local. This is kinda stupid.

LemonInTheDark commented on code/modules/atmospherics/machinery/air_alarm/_air_alarm.dm:
```diff
@@ -0,0 +1,529 @@
+
+	var/gas_dangerlevel = 0
+	for(var/gas_id in env_gases)
+		if(!(gas_id in cached_tlv)) // We're not interested in this gas, it seems.
+			continue
+		current_tlv = cached_tlv[gas_id]
```
```suggestion
		current_tlv = cached_tlv[gas_id]
		if(!current_tlv) // We're not interested in this gas, it seems.
			continue
```

LemonInTheDark replied:
also why current_tlv here? var creation overhead in theory exists but surely not important here

LemonInTheDark commented on code/modules/atmospherics/machinery/air_alarm/_air_alarm.dm:
```diff
@@ -0,0 +1,529 @@
+		if(!(gas_id in cached_tlv)) // We're not interested in this gas, it seems.
+			continue
+		current_tlv = cached_tlv[gas_id]
+		gas_dangerlevel = max(gas_dangerlevel, current_tlv.check_value(env_gases[gas_id][MOLES] * partial_pressure))
```
might be better to make the right side a second var, depends on readability in editor up to you

LemonInTheDark commented on code/modules/atmospherics/machinery/air_alarm/_air_alarm.dm:
```diff
@@ -0,0 +1,529 @@
+		current_tlv = cached_tlv[gas_id]
+		gas_dangerlevel = max(gas_dangerlevel, current_tlv.check_value(env_gases[gas_id][MOLES] * partial_pressure))
+
+	environment.garbage_collect()
```
why? we aren't asserting on it are we?

vincentiusvin replied:
this whole proc is oldcode ill rewrite it.

LemonInTheDark commented on code/modules/atmospherics/machinery/air_alarm/_air_alarm.dm:
```diff
@@ -0,0 +1,529 @@
+	//cache for sanic speed (lists are references anyways)
+	var/list/cached_tlv = tlv_collection
+
+	var/list/env_gases = environment.gases
```
move this above its loop, shouldn't stick it here (I get for the comment, hm s annoying)

LemonInTheDark commented on code/modules/atmospherics/machinery/air_alarm/_air_alarm.dm:
```diff
@@ -0,0 +1,529 @@
+		INVOKE_ASYNC(src, PROC_REF(apply_danger_level))
+	if(istype(selected_mode, /datum/air_alarm_mode/cycle) && environment_pressure < ONE_ATMOSPHERE * 0.05)
+		var/datum/air_alarm_mode/cycle/typed_mode = selected_mode
+		typed_mode.replace(my_area)
```
what is replaced for? why not have this logic inside the mode itself, in a base replaced() proc

vincentiusvin replied:
yeah this one kinda sucks ill figure something sane later.

LemonInTheDark commented on code/modules/atmospherics/machinery/air_alarm/_air_alarm.dm:
```diff
@@ -0,0 +1,529 @@
+/obj/machinery/airalarm/proc/apply_danger_level()
+
+	var/new_area_danger_level = 0
+	for(var/obj/machinery/airalarm/AA in my_area)
```
AA *cries*

LemonInTheDark commented on code/modules/atmospherics/machinery/air_alarm/_air_alarm.dm:
```diff
@@ -0,0 +1,529 @@
+		did_anything_happen = alarm_manager.clear_alarm(ALARM_ATMOS)
+	if(did_anything_happen) //if something actually changed
+		danger_level = new_area_danger_level
+
+	update_appearance()
```
update only if something changed? or is that comment wrong

LemonInTheDark commented on code/modules/atmospherics/machinery/air_alarm/_air_alarm.dm:
```diff
@@ -0,0 +1,529 @@
+
+MAPPING_DIRECTIONAL_HELPERS(/obj/machinery/airalarm, 24)
+
+/proc/extract_id_tags(list/objects)
+	var/list/tags = list()
+
+	for (var/obj/object as anything in objects)
+		tags += object.id_tag
+
+	return tags
+
+/proc/find_by_id_tag(list/objects, id_tag)
+	for (var/obj/object as anything in objects)
+		if (object.id_tag == id_tag)
+			return object
+
+	return null
```
should these be here? __HELPERS file maybe?

vincentiusvin replied:
We'll claim it as one of atmos'

LemonInTheDark reviewed (changes requested):
Reviewed the main air_alarm.dm file, will go through the rest... tomorrow?

vincentiusvin commented on code/modules/atmospherics/machinery/air_alarm/_air_alarm.dm:
```diff
@@ -0,0 +1,529 @@
+	var/datum/gas_mixture/environment = turf.return_air()
+	var/total_moles = environment.total_moles()
+	var/temp = environment.temperature
+	var/pressure = total_moles * R_IDEAL_GAS_EQUATION * temp / environment.volume
```
Did that on purpose so we dont iterate through the gaslist twice, but sure

LemonInTheDark replied:
Honestly man this is the coldest path imaginable. don't need to worry about it

vincentiusvin commented on code/modules/atmospherics/machinery/air_alarm/_air_alarm.dm:
```diff
@@ -0,0 +1,529 @@
+		var/portion = moles / total_moles
+		data["envData"] += list(list(
+			"name" = GLOB.meta_gas_info[gas_path][META_GAS_NAME],
+			"value" = "[round(moles, 0.01)] moles / [round(100 * portion, 0.01)] % / [round(portion * pressure, 0.01)] kPa",
```
Partial pressure because the thresholds go off them and this does the math for you
Ratio because the old one uses it.
Moles because it feels like it'd be useful.

Maybe we can afford to lose ratio if you want it trimmed.

Ill add the ui to the player body.

vincentiusvin commented on code/modules/atmospherics/machinery/air_alarm/_air_alarm.dm:
```diff
@@ -0,0 +1,529 @@
+				"path" = mode.type
+			))
+
+		// forgive me holy father
+		data["panicSiphonPath"] = /datum/air_alarm_mode/panic_siphon
+		data["filteringPath"] = /datum/air_alarm_mode/filtering
```
Yep

LemonInTheDark replied:
why not add a new act case and just do it on the backend?

vincentiusvin replied:
This is also used for the `selected` button prop in the UI since we send out the currently selected mode as a typepath. Doing that means we still need `isPanicSiphonSelected` alongside the ui_act stuffs.

Figured doing it this way would get us the smallest footprint.

vincentiusvin committed:
Reviews 1 and docs

vincentiusvin committed:
Webedit

Co-authored-by: LemonInTheDark <58055496+LemonInTheDark@users.noreply.github.com>

vincentiusvin committed:
Webedit 2

Co-authored-by: LemonInTheDark <58055496+LemonInTheDark@users.noreply.github.com>

vincentiusvin committed:
Simpler alarm system
The logic for multiple alarms in the same area should already be handled by the alarm manager, making the pooling redundant.
A bunch of other code rewrites to make it more modern and simple

vincentiusvin committed:
Fix webedit

vincentiusvin committed:
Overlay and light

vincentiusvin committed:
Moved

vincentiusvin committed:
Fix

vincentiusvin committed:
Init order memes

vincentiusvin commented:
Rewrote the alarm code a bit in [db70417](https://github.com/tgstation/tgstation/pull/72315/commits/db70417f9edeffbd95cb79d0ef5f23f4cf168a69), killed apply_danger_level off since it has weird behavior.

I think it's supposed to keep the area alarm going when our alarm becomes safe but there are still other unsafe alarms in the area, but that's not necessary since you already made the alarm_manager thing handle multiple source of alarms in one area.

Really weird paragraph isn't it, everything is named alarms.

jlsnow301 commented on tgui/packages/tgui/interfaces/AirAlarm.tsx:
```diff
@@ -0,0 +1,480 @@
+};
+
+const AirAlarmStatus = (props, context) => {
+  const { data } = useBackend<AirAlarmData>(context);
```
please destructure this

jlsnow301 commented on tgui/packages/tgui/interfaces/AirAlarm.tsx:
```diff
@@ -0,0 +1,480 @@
+    title: 'Alarm Thresholds',
+    component: () => AirAlarmControlThresholds,
+  },
+};
```
```suggestion
} as const;
```

For all and any global-ish constants.

Using `as const;` prevents it from being mutated, and it allows us to generate a type that we can use to constrain the available screens: `type Screen  = keyof typeof AIR_ALARM_ROUTES;`

jlsnow301 commented on tgui/packages/tgui/interfaces/AirAlarm.tsx:
```diff
@@ -0,0 +1,480 @@
+};
+
+const AirAlarmControl = (props, context) => {
+  const [screen, setScreen] = useLocalState(context, 'screen', 'home');
```
```suggestion
  const [screen] = useLocalState<Screen>(context, 'screen', 'home');
```

using the above type, constrain this to Screen 
we also don't actually need to get the setScreen function from localState unless we're using it

vincentiusvin replied:
![image](https://user-images.githubusercontent.com/54709710/210518126-e4e2407f-7d7c-4d58-a3cc-ea02380add4c.png)
![image](https://user-images.githubusercontent.com/54709710/210518146-be98f954-c823-4da4-976b-ee04b8170691.png)
![image](https://user-images.githubusercontent.com/54709710/210517864-f25d7bb3-1017-4ac4-a67c-c1b3b29b845c.png)

An element array doesn't have the props like key (since its an array) but is still a valid render. I can circumvent this by wrapping the arrays in a fragment, please let me know if this is bad.

jlsnow301 replied:
Wrapping the element array is fine, but I'm still wondering if using route.component() via direct invocation is the best way to do this. It may have no effect at all, but this is otherwise bad practice (a good question for @stylemistake)

https://dev.to/igor_bykov/react-calling-functional-components-as-functions-1d3l

If its possible, try setting the component property to `component: <Smth />`, then we should be able to use `const Component = route.component` or `{route.component}` in the return block

vincentiusvin replied:
Nah we didn't do direct invocation at all, route.component() returns a function, not an element.

jlsnow301 commented on tgui/packages/tgui/interfaces/AirAlarm.tsx:
```diff
@@ -0,0 +1,480 @@
+
+const AirAlarmControlHome = (props, context) => {
+  const { act, data } = useBackend<AirAlarmData>(context);
+  const [screen, setScreen] = useLocalState(context, 'screen', 'home');
```
same as above

vincentiusvin committed:
Frontend review

vincentiusvin committed:
Still suboptimal but its fine for now

Profakos commented:
Ever since this PR, air alarms runtime if a gas on a tile has 0 total_moles, could you please look into this? Linked issue: #72776
